spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/pubsub/health/PubSubSubscriptionHealthIndicatorAutoConfiguration.java

```diff
@@ -0,0 +1,96 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spring.autoconfigure.pubsub.health;
+
+
+import java.io.IOException;
+
+import com.google.api.gax.core.ExecutorProvider;
+import com.google.api.gax.core.FixedExecutorProvider;
+import com.google.cloud.monitoring.v3.MetricServiceClient;
+import com.google.cloud.spring.autoconfigure.pubsub.GcpPubSubAutoConfiguration;
+import com.google.cloud.spring.autoconfigure.pubsub.GcpPubSubProperties;
+import com.google.cloud.spring.pubsub.core.PubSubTemplate;
+import com.google.cloud.spring.pubsub.core.health.HealthTrackerRegistry;
+import com.google.cloud.spring.pubsub.core.health.HealthTrackerRegistryImpl;
+
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.boot.actuate.autoconfigure.health.CompositeHealthContributorConfiguration;
+import org.springframework.boot.actuate.autoconfigure.health.ConditionalOnEnabledHealthIndicator;
+import org.springframework.boot.actuate.health.HealthIndicator;
+import org.springframework.boot.autoconfigure.AutoConfigureBefore;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
+
+/**
+ * @author Emmanouil Gkatziouras
+ *
+ * @since 2.0.5
+ */
+@Configuration(proxyBeanMethods = false)
+@ConditionalOnClass({HealthIndicator.class, MetricServiceClient.class})
+@ConditionalOnEnabledHealthIndicator("pubsub-subscriber")
+@ConditionalOnProperty({"spring.cloud.gcp.pubsub.health.lagThreshold", "spring.cloud.gcp.pubsub.health.backlogThreshold"})
+@AutoConfigureBefore(GcpPubSubAutoConfiguration.class)
+@EnableConfigurationProperties(GcpPubSubProperties.class)
+public class PubSubSubscriptionHealthIndicatorAutoConfiguration  extends
+	CompositeHealthContributorConfiguration<PubSubHealthIndicator, PubSubTemplate> {
+
+	private final GcpPubSubProperties gcpPubSubProperties;
+
+	public PubSubSubscriptionHealthIndicatorAutoConfiguration(
+		GcpPubSubProperties gcpPubSubProperties) {
+		this.gcpPubSubProperties = gcpPubSubProperties;
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	public MetricServiceClient metricServiceClient() throws IOException {
+		return MetricServiceClient.create();
+	}
+
+	@Bean
+	@ConditionalOnMissingBean(name = "healthCheckExecutorProvider")
+	@ConditionalOnProperty("spring.cloud.gcp.pubsub.health.executorThreads")
+	public ExecutorProvider healthCheckExecutorProvider(
+			@Qualifier("pubsubSubscriberThreadPool") ThreadPoolTaskScheduler scheduler) {
```

Comment:
The healthcheck executor provider right now is still backed by the shared threadpool, which is not available at this point yet (since thisclass is autoconfigured before `GcpPubSubAutoconfiguration`.

Reply:
Added here, missed it previously

https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/7612d0c0296da4b032c6c0800f1dd5d5da64527c/spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/pubsub/health/PubSubSubscriptionHealthIndicatorAutoConfiguration.java#L76-L89

There is an obvious duplication on the Thread creation (`GcpPubSubAutoConfiguration`) so please advice if we create a helper function of we use another class.